### ISSUE TITLE
onSnapshot listener is being called multiple times unexpectedly

### ISSUE THREAD

**Issue by david-deepblocks** [firebase/firebase-js-sdk#1494]
## Environment

  * Operating System version: MacOS Mojave 10.4
  * Browser version: Chrome 71.0.3578.98
  * Firebase SDK version: 5.7.2
  * Firebase Product: database (auth, database, storage, etc)

## The problem

I seem to be having similar symptoms as explained in  [this](https://github.com/firebase/firebase-ios-sdk/issues/1045) other issue, except I'm using JS sdk. 
When I make a call to create a document, we attach a snapshot listener. Whenever this document changes we expect our listener to be called with a `isPendingWrites` flag set to `false`. If my understanding is correct, this means that a change has been made on the server side. Now, when creating/modifying this document, I expect this listener to be called **once** with the `isPendingWrites` flag set to `true`, thus ignoring it. The problem is, the listener is sometimes called more than once on write with that flag set to `false`,  unexpectedly.

#### Steps to reproduce:

I have a snapshot listener for a document which I'm creating and deleting frequently. When the document is created (the document and the collection don't exist in this scenario), the snapshot listener is being called multiple times with different and unexpected values. This happens sometimes when the document already exists and is being modified, or sometimes when the document is being created for the first time. This doesn't happen all the time, and I've tried so many things to recreate it that I can't seem to find concrete steps to recreate it. 
 
#### Relevant Code:
Attach Listener:
```
let lastSignInTime = new Date().toJSON();
let userSessionDocumentReference = firebase.firestore().collection("userSessions").doc(user.uid);
userSessionDocumentReference.set({ lastSignInTime: lastSignInTime })

sessionUnsubscribe = userSessionDocumentReference.onSnapshot(userSessionSnapshotListener);
```

Listener:
```
const snapShotListener = (snapshot) => {
 console.log(snapshot);
 if (!snapshot.metadata.hasPendingWrites) {
    //... logout
 }
} 
```

Code to delete entry:
```
try {
    await firebase.firestore().collection("userSessions").doc(user.uid).delete();
} catch (error) {
    log.error(error);
}
```

log of the snapshot object passed into the listener
```
exists=true
DocumentSnapshot {_firestore: Firestore, _key: DocumentKey, _document: Document, _fromCache: true, _hasPendingWrites: true}
exists=false
DocumentSnapshot {_firestore: Firestore, _key: DocumentKey, _document: null, _fromCache: false, _hasPendingWrites: false}
exists=true
DocumentSnapshot {_firestore: Firestore, _key: DocumentKey, _document: Document, _fromCache: false, _hasPendingWrites: false}
```

The first one is expected, but I don't know where the second and third come from. In the second snapshot the document doesn't exist, but on the third one it does. 

On log out, I'm calling the `sessionUnsubscribe()` to unsubscribe that listener. 
I'm also pretty certain I'm not attaching more than one listener.

Here is another log of login in and out of our platform which writes this document and adds a listener.
```
load platform --> authentication onstatechange: false 
sign in --> authentication onstatechange: true 
writes to document and then attach listener --> attached listener: ƒ () {  
            asyncObserver.mute();
            _this._firestoreClient.unlisten(internalListener);
        }
listener gets called on write as expected --> DocumentSnapshot {_firestore: Firestore, _key: DocumentKey, _document: Document, _fromCache: true, _hasPendingWrites: true} 
signing out and unsubscribing from listener --> authentication onstatechange: false 
sign in again --> authentication onstatechange: true
writes to document and then attach listener --> attached listener: ƒ () { 
            asyncObserver.mute();
            _this._firestoreClient.unlisten(internalListener);
        }
listener gets called on write as expected --> DocumentSnapshot {_firestore: Firestore, _key: DocumentKey, _document: Document, _fromCache: true, _hasPendingWrites: true} 
unexpected call to listener --> DocumentSnapshot {_firestore: Firestore, _key: DocumentKey, _document: null, _fromCache: false, _hasPendingWrites: false}
  exists: false
  id: (...)
  metadata: SnapshotMetadata
     fromCache: false
     hasPendingWrites: false
     __proto__: Object
  ref: (...)
  _document: null
  _firestore: Firestore {_queue: AsyncQueue, INTERNAL: {…}, _config: FirestoreConfig, _databaseId: 
  DatabaseId, _dataConverter: UserDataConverter, …}
  _fromCache: false
  _hasPendingWrites: false
  _key: DocumentKey {path: ResourcePath}
  __proto__: Object
```
I did log out and logged in multiple times before this happened. Then logged out and in multiple times after without getting the unexpected behaviour. It seems to happen at random.

**Comment by wilhuff** [comment]
Despite the details in here I'm still not sure what you're doing that demonstrates the problem. For example, you wrote:

> Code to delete entry

But immediately below you wrote:

> log of the snapshot object passed into the listener
> 
> ```
> exists=true
> DocumentSnapshot {_firestore: Firestore, _key: DocumentKey, _document: Document, _fromCache: true, _hasPendingWrites: true}
> ```

And then noted that you expected this first snapshot but based on the deletion you're doing I would expect the snapshot 

Your second log doesn't indicate where you're deleting the document at all. It almost seems as if you're not deleting the document and are surprised that the document is temporarily deleted but then I'm not sure how this relates to the code snippet you have for deleting the document.

Note that I'm not saying there's no problem here, just that I'm having trouble piecing together what the cause might be. 

The best way forward would be to come up with a small code snippet that demonstrates the problem, ideally as minimal as possible. For example, is the signing in and signing out required for the problem to occur?

A second path forward is to enable Firestore's logging with a call to [`Firestore.setLogLevel('debug')`](https://firebase.google.com/docs/reference/js/firebase.firestore.Firestore#setLogLevel). This will give us a log of what the SDK sent to the server and what it found locally etc. Changes derived from the change in sign-in state are also reflected in the log. If your logs are sensitive, feel free to send them to me directly at mcg@firebase.com.

**Comment by david-deepblocks** [comment]
@wilhuff thank you for responding so fast, and sorry for the lack of clarity.
[Here, I've set up a sample project that mimics our setup](https://stackblitz.com/edit/react-firebase-testing-zwkwfc?file=index.js). We're using React with redux. 

### Expected Behavior
Alert should only pop up when deleting the document.

### Current Behavior 
Alert shows up when creating the document. 

### Steps to reproduce
Create document, then delete document. Repeat multiple times. This is not consistent, sometimes the problem happens on the first try, sometimes it happens after 6 to 7 tries.

**Comment by mikelehen** [comment]
Thanks for the repro!  Taking a quick look, I think I may see the confusion:
```
if (!snapshot.metadata.hasPendingWrites) {
  alert("Document has been deleted!")
}
```

In general, `!snapshot.metadata.hasPendingWrites()` does not imply the document is deleted.  It just means you don't have any pending local writes against it (if you *do* have pending writes, it could be due to first creating the document, writing an update to the document, etc. not necessarily a delete).

When you create a document it's completely expected that you'll get an initial snapshot with `hasPendingWrites()` equal to true.  In fact, if you remove the `await` in your createDoc() function, then you'll *always* hit the alert.

Where the confusion is coming in I think is that there's a bit of an undefined ordering when you do a write as to whether the write Promise will complete first (i.e. your `await` in createDoc() will finish) or whether the listener will get notified that the write is complete (i.e. `hasPendingWrites()` will become false on the snapshot).  And so sometimes you see the alert and sometimes you won't.  This is a subtle consequence of the fact that reads and writes happen via independent streams to the backend under the covers and so you can't count on the ordering.

Does this help?  If you're really just trying to find out if the document has been deleted, you should do:

```
if (!snapshot.exists) {
  alert("Document has been deleted!")
}
```

If you're trying to accomplish something else, perhaps you can explain your use case and we may be able to provide more help.

**Comment by wilhuff** [comment]
Another thing I see in here is that you're adding your listener that examines metadata using onSnapshot but you're not passing the `includeMetadataChanges: true` option. This means that transitions between states @mikelehen is describing that don't otherwise change the data aren't raised. If you really are interested in `hasPendingWrites` make sure to include this option when listening.

**Comment by david-deepblocks** [comment]
Hey guys, thank you so much for all the support. Our apologies, yeah it looks like we had some issues in our code - specifically with missing `await` statements and the order in which we were attaching the listener and writing the document. We’ve fixed our issue, however, [here](https://stackblitz.com/edit/react-firebase-testing-zwkwfc?file=App.js) is example code that reproduces the strange listener behavior we were experiencing. even though we are no longer experiencing this, you may find it useful. 

To summarize the issue, I believe there might be some race conditions, as we were writing a document and then attaching our listener afterwards. It seemed to work as expected, except when sometimes that listener would call the callback unexpectedly. By attaching the listener first and adding an await on the write, we were able to fix our issue. 

Again, thank you for your time.

**Comment by wilhuff** [comment]
I'm glad to hear you got past this.

In general the incremental cost of a listener beyond the first is next to nothing, so you're much better off adopting a strategy where you create listeners that match the lifetimes of your components and avoid creating and removing them based on when you expect updates to be happening. This will allow you to disable listeners when components become inactive without impeding your ability to make changes your components might otherwise be interested in.

In any case, your example code is still making the wrong inference:

```js
if (!snapshot.metadata.hasPendingWrites) {
  alert("Document has been deleted!")
}
```

`hasPendingWrites` has nothing to do with whether or not a document is deleted. When `hasPendingWrites` is true, we're indicating that the client has a write in its outbound queue.

The race condition you're observing is that sometimes the client is managing to commit the new document before the listener starts. At that point there are no pending writes to the document so `hasPendingWrites` is false.

Use `snapshot.exists` to determine if a document exists or not. Use `snapshot.metadata.hasPendingWrites` (with `includeMetadataChanges: true` on the listener) to detect whether or not the document has unsynchronized changes pending (for example to show a saving spinner or similar indicator).
